Health Endpoint Design Document (#2503)
Azure/data-api-builder: docs/design/HealthEndpoint.md

# Design Document: DAB Health Endpoint
## Objective:

> The objective of this task is to enhance the "Health Endpoint" for DAB. Currently we only show whether the DAB Engine is healthy or not and which version of DAB is running. However, the objective under this task item is to enhance this to support delatiled description of each supported (enabled) endpoint with all different kinds of parameters that would be defined below.

Reviewer: to support each endpoint? do you mean for each entity?

Author: Endpoint here refers to both entity and datasource as we need to show the health of both of these endpoints. Updating to clearly state the same in document

> The objective of this task is to enhance the "Health Endpoint" for DAB. Currently we only show whether the DAB Engine is healthy or not and which version of DAB is running. However, the objective under this task item is to enhance this to support delatiled description of each supported (enabled) endpoint with all different kinds of parameters that would be defined below.
> 

Reviewer 2: Should be `detailed` instead of `delatiled`
```suggestion
The objective of this task is to enhance the "Health Endpoint" for DAB. Currently we only show whether the DAB Engine is healthy or not and which version of DAB is running. However, the objective under this task item is to enhance this to support detailed description of each supported (enabled) endpoint with all different kinds of parameters that would be defined below.

```

> For Example, if the data source health is being performed and the user mentions that we would do a GET of all resources from the DB and this should be under 25sec. This check would be given as a SQL Query that the users wants to execute on the DB, then after performing this check, we will validate if the condition is satisfied, else this data source moniker would be considered unhealthy.

## Implementation Details
### Schema
Health check responses follow a common convention rather than a strict standard. The typical pattern involves a "checks" property for individual components' statuses (e.g., database, memory), with each status rolling up to an overall "status" at the top level.
```
{
  "status": "Healthy",
  "checks": {
    "check-name": { "status": "Healthy" },
    "check-name": { "status": "Healthy" }
  }
}
```

### Configuration Updates
We need to update the dab config file to include details of the health check for different configuration parameters like runtime, data source, entities. 

#### `runtime.health` Configuration

> The runtime configuration would include details like cache-ttl in case we need to cache the response of health checks, the max-drop value which tells you the degree of parallelism i.e. how many queries that DAB shoud run at once to get health results and roles i.e. which role is allowed to view the health information of DAB.

Reviewer 3: ```suggestion
The runtime configuration would include details like cache-ttl in case we need to cache the response of health checks, the max-dop value which tells you the degree of parallelism i.e. how many queries that DAB shoud run at once to get health results and roles i.e. which role is allowed to view the health information of DAB.
```

| **Property**   | **Data Type** | **Required** | **Default** | **Description**                                                                                      |
|----------------|---------------|--------------|-------------|------------------------------------------------------------------------------------------------------|
| `enabled`     | Boolean       | No           | `true`      | Enables or disables health checks at the runtime level.                                             |

> | `cache-ttl`   | Integer       | No           | `5`         | Time-to-live (in seconds) for caching health check results.                                          |

Reviewer 3: mention the unit of sec in the name of the property for eg: `cache-ttl-sec`

| `max-dop`     | Integer       | No           | `1`         | Maximum Degree of Parallelism for running health checks.                                             |

> | `roles`       | Array         | No           | `*`         | Roles allowed to access the health endpoint (e.g., `anonymous`, `authenticated`).                   |

Author: * would match with all the items in the list. Either authenticated or anonymous

| **Property**      | **Data Type** | **Required** | **Default** | **Description**                                                                                      |
|-------------------|---------------|--------------|-------------|------------------------------------------------------------------------------------------------------|
| `moniker`         | String        | No           | `NULL`      | Identifier for the data source; useful when multiple data sources exist.                            |
| `enabled`         | Boolean       | No           | `true`      | Enables or disables health checks for the data source.                                              |
| `query`           | String        | No           | N/A         | Custom SQL query used to perform the health check.                                                  |
| `threshold-ms`    | Integer       | No           | `10000`     | Threshold in milliseconds for the query response time before the check is considered degraded.      |

#### `<entity-name>.health` Configuration

> The Entity config parameters contain information about the GET filter or query which needs to be carried out on the data source entiy and under what threshold should the response be received for it to qualify as a healthy entity.

Reviewer 2: Should be `entity`
```suggestion
The Entity config parameters contain information about the GET filter or query which needs to be carried out on the data source entity and under what threshold should the response be received for it to qualify as a healthy entity.
```

| **Property**     | **Data Type** | **Required** | **Default** | **Description**                                                                                      |
|------------------|---------------|--------------|-------------|------------------------------------------------------------------------------------------------------|
| `enabled`        | Boolean       | No           | `true`      | Enables or disables health checks for the specific entity.                                          |

> | `filter`         | String        | No           | `null`      | Filter condition applied to the health check query (e.g., `"Id eq 1"`).                             |

Reviewer 2: Why would we need a filter for the health check?

Author: Filter is to create a query for entities and execute it that specific filter to get the response time

| `first`          | Integer       | No           | `1`         | Number of records to query during the health check.                                                 |
| `threshold-ms`   | Integer       | No           | `10000`     | Threshold in milliseconds for the query response time before the check is considered degraded.      |

#### Example
```
{
  "runtime" : {
    "health" : {
      "enabled": true, (default: true)
      "cache-ttl": 5, (optional default: 5)
      "max-dop": 5, (optional default: 1)
      "roles": ["anonymous", "authenticated"] (optional default: *)
    }
  }
}
{
  "data-source" : {
    "health" : {

>       "moniker": "sqlserver", (optional default: NULL) // not required, on purpose, most have just one

Reviewer: what is `moniker`?

      "enabled": true, (default: true)
      "query": "SELECT TOP 1 1", (option)

>       "threshold-ms": 100 (optional default: 10000)

Reviewer 2: I think it might be good to show how it would look like when there are multiple `data-source`

Author: This would be handled later, in another issue.

    }
  }
}
{
  "<entity-name>": {
      "health": {
        "enabled": true, (default: true)
        "filter": "Id eq 1" (optional default: null),
        "first": 1 (optional default: 1),
        "threshold-ms": 100 (optional default: 10000)
      },
      ...
    },
  }
}
```

> The idea of using this updates configuration is to allow the developer to influence how the checks work against the datasource/entity. This would provide him with a more detailed process for checking if DAB engine is healthy and would give him an enhanced user experience. 

Reviewer 3: ```suggestion
The idea of using this updates configuration is to allow the developer to influence how the checks work against the datasource/entity. This would provide them with a more detailed process for checking if DAB engine is healthy and would give them an enhanced user experience. 
```
Being more neutral on the use of pronouns.

### Output Sample
```
{
  "status": "Unhealthy",

>   "status": "Healthy",

Reviewer 2: Why are there 2 status?

      "/App/dab-config.json ({data-source-moniker})",
      "/App/dab-config-2.json ({data-source-moniker})"
    ],
    "dab-schemas": [
      "/App/schema.json"
    ]
  },
  "checks": {
    "database-moniker" : {
        "status": "Healthy",
        "tags": ["database", "performance"]
        "description": "Checks if the database is responding within an acceptable timeframe.",
        "data": {
            "responseTimeMs": 10,
            "maxAllowedResponseTimeMs": 10
        }
    },
    "database-moniker" : {
        "status": "Unhealthy",
        "tags": ["database", "performance"],
        "description": "Checks if the database is responding within an acceptable timeframe.",
        "data": {
            "responseTimeMs": 20,
            "maxAllowedResponseTimeMs": 10
        }
    },
    "database-moniker" : {
        "status": "Unhealthy",
        "tags": ["database", "performance"]
        "description": "Checks if the database is responding within an acceptable timeframe.",
        "data": { 
            "responseTimeMs": NULL,
            "maxAllowedResponseTimeMs": 10
        },
        "exception": "TimeoutException: Database query timed out."
    },

>     "<entity-name>": {

Author: If there is a configuration issue, and the engine has not started.. then the health endpoint will also not be displayed. Health is displayed when the engine is started and we are able to hit localhost:500. This endpoint would contain information of each entity .. this information signifies whether basic query like `get all` is running smoothly and under the threshold timelimit or not. I dont see any issue in case entity is ill configured, but lets discuss more in detail

>     "<entity-name>": {

Reviewer: what about db issues? are we assuming that if the engine is started all the endpoints would be healthy? if not, how will we be knowing status of each endpoint?

>     "<entity-name>": {

Author: No, Health is checked only when the user tries to access the health endpoint. Hence, only when he hits the endpoint the queries are called. 
In case there are db issues, the health would be "unhealthy"

| **Term**           | **Description**                                         |
|--------------------|---------------------------------------------------------|
| **Health Endpoint** | The URL (e.g., `/`) exposed as JSON.             |
| **Check**           | A specific diagnostic test (e.g., database, API).      |
| **Status**          | The result of a check.                                 |
| **Status.Healthy**  | The system is functioning correctly.                   |
| **Status.Unhealthy**| The system has a critical failure or issue.            |

## Proposed Change
We want to create a detailed version of the DAB Engine Health endpoint with information regarding all REST and GraphQL endpoints and their behavior. Here we would update the configuration schema to include what all results or checks do we want to perform on that data-source/entity (described below) to validate if they are healthy.
We want the user to provide a check and threshold value what DAB would carry out and then check if the result is under the threshold, if so, the DAB Engine would be considered healthy for that endpoint, else unhealthy. 

**Case 1: Datasource**\

> Here we would carry out standard queries based on the datasource type and execute them under the given threshold. If the engine gets the result below the specified threshold, the DB is considered healthy else unhealthy.

Reviewer: It would be good to add those standard queries that is being mentioned here.

> > Example: If we have a SQL Server, we would execute the standard query given belowe for SQL Server and if the elapsed time for this query is under the threshold, the data source would be considered healthy.

Reviewer: ```suggestion
> Example: If we have a SQL Server, we would execute the standard query given below for SQL Server and if the elapsed time for this query is under the threshold, the data source would be considered healthy.
```

This would form a query to REST and GraphQL endpoints to check their health.

> Example: In case for entity `UserTable` we say that we need to run a health query with first = 5 under 100ms threshold. While running the DAB Engine we would execute the REST endpoint for `UserTable` and fetch the fist 5 rows and similar for GraphQL and check if the threshold is under 100ms. If so, it would be consider healthy entity.

## Implementation Details
### Schema
Health check responses follow a common convention rather than a strict standard. The typical pattern involves a "checks" property for individual components' statuses (e.g., database, memory), with each status rolling up to an overall "status" at the top level.
```
{
  "status": "Healthy",
  "checks": {
    "check-name": { "status": "Healthy" },
    "check-name": { "status": "Healthy" }
  }
}
```

### Configuration Updates
We need to update the dab config file to include details of the health check for different configuration parameters like runtime, data source, entities. 

#### `runtime.health` Configuration

> The runtime configuration would include details like cache-ttl in case we need to cache the response of health checks, the max-dop value which tells you the degree of parallelism i.e. how many queries that DAB should run at once to get health results and roles i.e. which role is allowed to view the health information of DAB.

Reviewer: ```suggestion
The runtime configuration would include details like cache-ttl in case we need to cache the response of health checks, the max-dop value which specifies the degree of parallelism i.e. how many queries that DAB should run at once to get health results and roles i.e. which role is allowed to view the health information of DAB.
```

| `enabled`        | Boolean       | No           | `false`      | Enables or disables health checks for the specific entity.                                          |
| `first`          | Integer       | No           | `1`         | Number of records to query during the health check. Using this value the GET query would be created to fetch records for REST and GraphQL endpoints.                                                 |
| `threshold-ms`   | Integer       | No           | `10000`     | Threshold in milliseconds for the query response time before the check is considered degraded.      |

#### Example
```
{
  "runtime" : {
    ...
    "health" : {
      "enabled": true, (default: false)
      "cache-ttl": 5, (optional: 5)
      "max-dop": 5, (optional default: 1)
      "roles": ["anonymous", "authenticated"] (required)
    }
  }
}
{
  "data-source" : {
    ...
    "health" : {
      "name": "mssql", (optional default: Database Type) // not required as mostly configs have just one
      "enabled": true, (default: false)

>       "query": "SELECT TOP 1 1", (Based on DB Type)

Reviewer: i thought we decided to remove this query from here and not allow use to configure.

Author: yes based on DB Type we will take it up

#### Rest Query
We take the base URL which is http://localhost:5000 and append the REST path in the suffix. After this we get http://localhost:5000/api.\
For each entity we have rest as a parameter in the config block as well in case the path of entity is different from its name. We build the entity REST path using `<entity-name>.rest.path ?? entityName`. This means if the path is given use that, else the entity name is the path to append to suffix.
Further our call becomes ` http://localhost:5000/api/usertable`.

Finally we add the query parameters which is first to get the query to execute as `http://localhost:5000/api/usertable?$first=5`. This query is then executed to get the elapsed time and we get whether the REST endpoint of UserTable healthy or not.

#### GraphQL Query

To execute this query we first need to get the schema of that particular table. Hence for each table we run the `introspection query` to get the schema which gives the column names which are then used to query the GraphQL endpoint to check the health.

> For this we run a **POST** query against the base URL **http://localhost:5000/graphl/** with a Request BODY.\

Reviewer: ```suggestion
For this we run a **POST** query against the base URL **http://localhost:5000/graphql/** with a Request BODY.\
```

**GraphQL Query**\
After getting column names for the entity we create the graphQL query payload 
`
 query = $"{{{entityName.ToLowerInvariant()}(first: {First}) {{items {{ {string.Join(" ", columnNames)} }}}}}}"
`. We execute a POST query against the GraphQL base URL.\
CURL Command
```
curl --request POST \
  --url http://localhost:5000/graphql \
  --header 'Content-Type: application/json' \
  --header 'User-Agent: insomnia/10.3.0' \
  --data '{
	"query": "{reviews(first: 4) {items { content id }}}"
}'
```
The time it took to execute the above query is the time elapsed for checking the graphQL health.

## Output Sample
```
{

>   "status": "Unhealthy",/"status": "Healthy",

Reviewer: ```suggestion
  "status": "Unhealthy/Healthy",
```

  "version": "1.2.10",
  "app-name": "dab_oss_1.2.10",
  "dab-configuration": {
    "http": true,
    "https": true,
    "rest": true,
    "graphql": true,
    "telemetry": true,
    "caching": true,
    "mode": "development"
  },
  "checks": {
    "database-name" : {
        "status": "Healthy",
        "tags": ["database", "performance"]
        "description": "Checks if the database is responding within an acceptable timeframe.",
        "data": {
            "responseTimeMs": 10,
            "maxAllowedResponseTimeMs": 10
        }
    },
    "<entity-name>": {
      "status": "Healthy",
      "description": "Checks if the endpoint is responding within an acceptable timeframe.",
      "tags": ["endpoint", "REST"]
      "data": {
          "responseTimeMs": 10,
          "maxAllowedResponseTimeMs": 10
      }
   },

>     "<entity-name>": {

Reviewer: for REST and GRAPHQL will there be separate JSON objects?

Author: yes we would have different objects for rest and graphql

      "status": "Unhealthy",
      "description": "Checks if the endpoint is responding within an acceptable timeframe.",
      "tags": ["endpoint", "GRAPHQL"]
      "data": {
          "responseTimeMs": 20,
          "maxAllowedResponseTimeMs": 10
      }
   },
    "<entity-name>": {
      "status": "Unhealthy",
      "description": "Checks if the endpoint is responding within an acceptable timeframe.",
      "tags": ["endpoint", "REST"]
      "data": {
          "responseTimeMs": 20,
          "maxAllowedResponseTimeMs": 10
      }
      "exception": "{exception-message-here}"
   },
  }
}
```

## Limitations

> + We do not support health checks for stored procedures.

Reviewer: I think we are also not supporting hot-reload right? it would be good to add as limitation as well.

Author: added

# Design Document: DAB Health Endpoint

> ## Objective:

Reviewer 3: PR description should remove the template description and provide details specific to this PR.

The objective of this task is to enhance the "Health Endpoint" for DAB. Currently we only show whether the DAB Engine is healthy or not and which version of DAB is running. However, the objective under this task item is to enhance this to support detailed description of each supported (enabled) endpoint with all different kinds of parameters that would be defined below.

> The objective of this task is to enhance the "Health Endpoint" for DAB. Currently we only show whether the DAB Engine is healthy or not and which version of DAB is running. However, the objective under this task item is to enhance this to support detailed description of each supported (enabled) endpoint with all different kinds of parameters that would be defined below.

Reviewer 3: Repetitive text on line 4 - same as line 3.

Reviewer 3: I still the repeated lines. Were they really removed? Perhaps only in your local copy?
Please push your changes to the remote before marking the comment as resolved. Unresolving the conversation for now.

Author: Please check the latest commit. It is corrected
<img width="892" alt="image" src="https://github.com/user-attachments/assets/7c4087ee-fca2-4725-9f46-ba06bd3771e4" />

```
{
  "status": "Unhealthy/Healthy",
  "version": "1.2.10",
  "app-name": "dab_oss_1.2.10",
  "dab-configuration": {
    "http": true,
    "https": true,
    "rest": true,
    "graphql": true,
    "telemetry": true,
    "caching": true,
    "mode": "development"
  },
  "checks": {
    {
      "name": "database-name",
      "status": "Healthy",
      "tags": ["data-source", "mssql"],
      "data": {
          "responseTimeMs": 10,
          "maxAllowedResponseTimeMs": 10
      }
    },
    {
      "name": "<entity-name>",
      "status": "Healthy",
      "tags": ["endpoint", "rest"],
      "data": {
          "responseTimeMs": 10,
          "maxAllowedResponseTimeMs": 10
      }
    },
    {
      "name": "<entity-name>",
      "status": "Healthy",
      "tags": ["endpoint", "graphql"]
      "data": {
          "responseTimeMs": 20,
          "maxAllowedResponseTimeMs": 50
      }
    },
    {
      "name": "<entity-name>",
      "status": "Unhealthy",
      "tags": ["endpoint", "graphql"]
      "exception": "{exception-message-here}",
      "data": {
          "responseTimeMs": 20,
          "maxAllowedResponseTimeMs": 10
      }
    },
  }
}
```

## Test Scenarios

> Health Check result scenarios for different cases of Health check and GraphQL and Rest entities enabled or disabled.\

Author: This is needed to insert a new line

'_' means enabled/disabled (doesn't matter)

* Global Health Enabled
  * Global GraphQL Enabled
    * Entity health ENABLED and Entity GraphQL ENABLED : Health is shown for this particular entity (Positive Scenario)
    * Entity health DISABLED and Entity GraphQL ENABLED : No GraphQL Health check for this particular entity

>     * Entity health _ and Entity GraphQL DISABLED : Validation Error in Config

Reviewer: what does `_` means? why validation error in config?

>     * Entity health _ and Entity GraphQL DISABLED : Validation Error in Config

Reviewer: why validation error if graphQL is disabled at entity level?

Author: This is the behavior without the healthEndpoint as well. havent modified anything specific related to this

Reviewer: I verified with the latest stable release and i don't see this case happening.

https://github.com/user-attachments/assets/066b451b-3098-4c64-93ee-57f7da0b2cc0

I also verified with the `dab validate`.
![image](https://github.com/user-attachments/assets/7d7a1f63-b659-4b9e-b639-591d7f1e88e6)

Also, if it's happening, please verify once with the team. It's not the expected behaviour for the health check as I mentioned in the demo as well.

Author: I have verified on my local in the main branch which has no health changes by putting the global graphQL as enabled and entity graphql for Publishers as disabled. 
I am still seeing the validate issues. Please check 
<img width="740" alt="image" src="https://github.com/user-attachments/assets/e2fa4ae6-d08e-49c0-bf9e-2273e999d8ad" />

Attached the config which I am using as well
Config
```
  "runtime": {
    "rest": {
      "enabled": true,
      "path": "/api",
      "request-body-strict": true
    },
    "graphql": {
      **"enabled": true,**
      "path": "/graphql",
      "allow-introspection": true,
      "multiple-mutations": {
        "create": {
          "enabled": true
        }
      }
    },
    "host": {
      "cors": {
        "origins": [
          "http://localhost:5000"
        ],
        "allow-credentials": false
      },
      "authentication": {
        "provider": "StaticWebApps"
      },
      "mode": "development"
    }
  },
  "entities": {
    "Publisher": {
      "source": {
        "object": "publishers",
        "type": "table"
      },
      "graphql": {
        **"enabled": false,**
        "type": {
          "singular": "Publisher",
          "plural": "Publishers"
        }
      },
      "rest": {
        "enabled": true
      },
```

Author: I tried the same thing via the latest stable release as well.
 dab validate -c .\dab-config.MsSql.json
Information: Microsoft.DataApiBuilder 1.3.19
<img width="599" alt="image" src="https://github.com/user-attachments/assets/1ea8b6a6-60be-43e9-b02d-061a878f8ad0" />

<img width="636" alt="image" src="https://github.com/user-attachments/assets/41213461-e63c-4a1d-8b22-0f2dea90cf46" />

Reviewer: I see that you see this error happening, but it's not the expected behavior. Instead of documenting it as the expected outcome, we should address the issue itself. The design doc should describe the intended behavior, so that if something doesn't work as expected in the future, we can refer back to it and make the necessary fixes.

Author: yes I have multiple entries and even after clearing the cache I am seeing the same result as before. If this is not expected then I will debug further as to why its occuring however i have modified the design docuemnt with the expected one

# Design Document: DAB Health Endpoint
## Objective:
The objective of this task is to enhance the "Health Endpoint" for DAB. Currently we only show whether the DAB Engine is healthy or not with a small description of the version.\
However, the objective under this task item is to enhance the health endpoint to support detailed description of each enabled entity by performing health checks on them based on different kinds of parameters that would be defined below.

## Need for DAB Health Check Endpoint
Azure App Service & Azure Kubernetes Service (AKS) support health probes to monitor the health of your application. If a service fails health checks, Azure can automatically restart it or redirect traffic to healthy instances.\
Similarly, we need Health Endpoint for Data API builder because if it fails health checks in a way a customer deems past a threshold, they have the option to recycle the container or send an alert to direct engineers.

## Current Setup

> There is no official industry standard for the health endpoint. /health or variations like /_health are common by convention.\

Reviewer 4: Could you more clearly state that this feature will have two endpoints? / and /health. That / (the root) returns what it already returns today, the basic health) and that /health returns the more comprehensive health depending on a mixture of if health is enabled and the user meets the roles requirement?

```
{
  "status": "Healthy",
  "checks": {
    "check-name": { "status": "Healthy" },
    "check-name": { "status": "Healthy" }
  }
}
```

### Configuration Updates
We need to update the dab config file to include details of the health check for different configuration parameters like runtime, data source, entities. 

#### `runtime.health` Configuration

The runtime configuration would include details like `cache-ttl-sec` in case we need to cache the response of health checks, the `max-dop` value which specifies the degree of parallelism i.e. how many queries that DAB should run at once to get health results and `roles` i.e. which role is allowed to view the health information of DAB.

| **Property**   | **Data Type** | **Required** | **Default** | **Description**                                                                                      |
|----------------|---------------|--------------|-------------|------------------------------------------------------------------------------------------------------|

> | `enabled`     | Boolean       | No           | `false`      | Enables or disables the comprehensive health checks for DAB Engine. In case of disabled; we show the previous format of health check.                                             |

Author: updated

| `cache-ttl-sec`   | Integer       | No           | `5`         | Time-to-live (in seconds) for caching health check results. If this value is not specified, caching would not be enabled. Currently, caching is not implemented hence, value is considered to be null.                                        |
| `max-dop`     | Integer       | No           | `1`         | Maximum Degree of Parallelism for running health checks.                                             |

> | `roles`       | Array of strings         | Yes           | NA         | Roles allowed to access the health endpoint (e.g., `anonymous`, `authenticated`).                   |

Author: Added

#### `data-source.health` Configuration

The database type in the data source health config determines the query that we should run on the data source. The time to run this query should come under the user specified threshold of ms to qualify as a healthy data source for DAB. We get the database type from the runtime parameters to get the query to run on the specific DB Type.

 > TODO: Handle Health Endpoint for multiple data-source configs in the upcoming enhancements

| **Property**      | **Data Type** | **Required** | **Default** | **Description**                                                                                      |
|-------------------|---------------|--------------|-------------|------------------------------------------------------------------------------------------------------|
| `enabled`         | Boolean       | No           | `true`      | Enables or disables health checks for the data source.                                              |

> | `name`         | Array of String        | No (Yes in case of multiple data source)           | Database Type      | Identifier for the data source; useful when multiple data sources exist.  

Reviewer 4: By the way, we DO support multiple data sources.

#### Example
```
{
  "runtime" : {
    ...
    "health" : {
      "enabled": true, (default: false)
      "cache-ttl-sec": 5, (optional; default: null) // Default value would be updated to 5sec once caching is enabled
      "max-dop": 5, (optional; default: 1)
      "roles": ["anonymous", "authenticated"] (required)
    }
  }
}
{
  "data-source" : {
    ...
    "health" : {
      "name": ["mssql"], (optional; default: Database Type) // not required as mostly configs have just one
      "enabled": true, (default: false)
      "threshold-ms": 100 (optional; default: 10000)
    }
  }
}
{
  "<entity-name>": {
      "health": {
        "enabled": true, (default: true)
        "first": 1 (optional; default: 1),
        "threshold-ms": 100 (optional; default: 10000)
      },
      ...
    },
  }
}
```
The idea of using this updated configuration is to allow the developer to influence how the health checks work against the datasource/entity. This would provide them with a more detailed process for checking if DAB engine is healthy and would give them an enhanced user experience. 

> ## Code Details

Reviewer 4: It's almost like there are two documents here, each repeating the other when they should be mixed together so you do not repeat yourself so much - this will lead us to get details out of sync.

Author: There are two sections in the document. 
First one is referring to the config changes which includes runtime, data-source and entities.
Second section includes the code details about how these above config parameters are being used to calculate the health config.

In case `runtime.health.enabled` is false/null then we show the user the original health endpoint format at both `'/'` and `'/health'`.\
In case `runtime.health.enabled` is true, then the comprehensive health report is displayed based on the rules and check below.

### Roles

> We focus on two aspects in terms of roles for health report. 

Author: The concept of runtime mode has not been discussed yet. Lets not make changes suddenly without discussing it in detail. We can connect over a call once, after which I am happy to incorporate this

> We focus on two aspects in terms of roles for health report. 

Author: incorporated

> + First is to check the roles present in the `runtime.health.roles` array in the config. We check if the incoming user has access to the comprehensive health check report i.e. they are in this array of allowed roles. For Health report of DB, we only need to check the first condition. 

Reviewer 4: It is strange that you discuss the same thing in two places - seems risky to get out of sync.

> + First is to check the roles present in the `runtime.health.roles` array in the config. We check if the incoming user has access to the comprehensive health check report i.e. they are in this array of allowed roles. For Health report of DB, we only need to check the first condition. 

Author: First section is 'what options the user can input' and second is after the user input 'how we interpret it and use it to calculate health'. 

Will include this role dependence on runtime after discussion.

+ Postgres: SELECT 1 LIMIT 1;
+ MySQL: SELECT 1 LIMIT 1;
+ Cosmos DB: SELECT TOP 1 1 FROM c
+ MS SQL: SELECT TOP 1 1

### Entity Health Check

In the config, for each entity we have three health parameters (Enabled, First and Threshold) 

+ Enabled: This would define whether we need to run the comprehensive health check for this role
+ First: This would be used to create the query for endpoints to fetch first 'x' records for the DB.
+ Threshold: This is used to check the time elapsed to run the above query.

To form the query we run a for loop against the entities in the config file. For each entity, we add two parameters in the checks array of the output. One for REST and second for GraphQL, these calls are differentiated using the Tags array in the output report.\

While running the health checks for entities, we need the BASE URL on which DAB is running which we use to run all rest and graphql queries on our engine. 
> Note: Current implementation we consider `http://localhost:5000` as the base URL. Need to track this, in case different. Everywhere in the document, using `baseUrl` in place of `http://localhost:5000`.

> #### Rest Query

Author: yes that is correct.  Please take the test scenarios below

We take the base URL and append the REST path in the suffix. After this we get baseUrl/api. (Default Rest Path: /api)\
For each entity we have rest as a parameter in the config block as well in case the path of entity is different from its name. We build the entity REST path using `<entity-name>.rest.path ?? entityName`. This means if the path is given use that, else the entity name is the path to append to suffix.
Further our call becomes ` baseUrl/api/usertable`. (Assuming the entity name is UserTable)

Finally we add the query parameters which is the first 'x' values to get the query as `baseUrl/api/usertable?$first=5`. This query is then executed to get the elapsed time and we get whether the REST endpoint of UserTable healthy or not.

> #### GraphQL Query

Author: yes that is correct. Please take the test scenarios below

}'

> ```

Reviewer 4: I am not sure I understand. Do you really use CURL for this check?

Author: yes i use this to get column Information.

The above introspection query is formatted at `src/media/introspection-query.graphql`.

> **Introspection Query Filter**

Reviewer 4: Why are you not using the metadata of the entity to build this? The GraphQL schema is built from the entity metadata and you already have the entity metadata, wouldn't that save you this costly step?

Author: Not sure if I understand you correctly. Lets discuss on call

Author: done

This gives us the GraphQL Schema which is deserialized to get the column names.\
Now we need to create a `columnNames` array which contains the list of column names which can be used to create the graphQL query.\
In the output of the above CURL command we get `Data.Schema.Types` array which contains these column names. We loop through this array and match the `type.kind` value with `OBJECT` and `type.name` value with the name of the entity i.e. `UserTable`. (Note: Only one entry satisfies this check)
After identifying that specific entity `type` that satisfies the above conditions, we get the object that contains the entity column names. However an entity can have primitive column types or Object column type which are nested further. These object nested types cannot be used to create the graphql query and hence we need to filter out the primitive data types. To resolve this we loop on `type.fields` array and match it with two conditions. If either of the conditions match, we add it to the column names array.\
For each field we get the `field.type` object 
+ Condition 1: If the `field.type.kind` is `SCALAR` then this is a primitive column name.

> + Condition 2: If the `field.type.kind` is `NON_NULL` then check if `field.type.ofType.kind` is `SCALAR`. 

Reviewer 4: All of these rules would not be necessary if you used the entity metadata to infer the GraphQL query instead of using introspection. This feels needlessly complex and brittle. Since you already have the entity metadata and since we know the GraphQL schema is built from the metadata, you should be able to more efficiently infer the GraphQL query from the metadata than rely on this process and bag of extra rules you are listing here.

Author: Duplicate comment

We add all those column names which satisfy either of these two conditions to the top `columnNames` array and finally use this to create the graphQL query.

**GraphQL Query**\
After getting column names for the entity we create the graphQL query payload 
`query = $"{{{entityName.ToLowerInvariant()}(first: {First}) {{items {{ {string.Join(" ", columnNames)} }}}}}}"
`. We execute a **POST** query against the GraphQL base URL.\
CURL Command
```
curl --request POST \
  --url baseUrl/graphql \
  --header 'Content-Type: application/json' \
  --header 'User-Agent: insomnia/10.3.0' \
  --data '{
	"query": "{UserTable(first: 4) {items { content id }}}"
}'

> ```

Reviewer 4: This would be easier to read a C#:

```csharp
using System.Text.Json;
using System.Net.Http;
using System.Diagnostics;
using System.Threading;

var queryText = "{UserTable(first: 4) {items { content id }}}";

async Task<TimeSpan> ExecuteAsync(string query, CancellationToken cancellationToken = default)
{
    var httpClient = new HttpClient();
    var queryUrl = "baseUrl/graphql";

    var queryObject = new { query = query };
    var queryJson = JsonSerializer.Serialize(queryObject);
    var queryContent = new StringContent(queryJson, System.Text.Encoding.UTF8, "application/json");

    var stopwatch = new Stopwatch();

    stopwatch.Start();
    var responseObject = await httpClient.PostAsync(queryUrl, queryContent, cancellationToken);
    stopwatch.Stop();
    
    responseObject.EnsureSuccessStatusCode();
    var responseText= await responseObject.Content.ReadAsStringAsync();
    var responseJson = JsonDocument.Parse(responseText);

    return TimeSpan.FromMilliseconds(stopwatch.ElapsedMilliseconds);
}
```

Author: I am unable to understand why we would include code details in the design doc. CURL command gives all the information the coder would need to create the above snippet. 
Can you elaborate why you feel this is required?

```
{
  "status": "Unhealthy/Healthy",
  "version": "1.2.10",
  "app-name": "dab_oss_1.2.10",
  "dab-configuration": {
    "http": true,
    "https": true,
    "rest": true,
    "graphql": true,
    "telemetry": true,
    "caching": true,
    "mode": "development"
  },
  "checks": {
    {
      "name": "database-name",
      "status": "Healthy",
      "tags": ["data-source", "mssql"],
      "data": {
          "responseTimeMs": 10,
          "maxAllowedResponseTimeMs": 10
      }
    },
    {
      "name": "<entity-name>",
      "status": "Healthy",
      "tags": ["endpoint", "rest"],
      "data": {
          "responseTimeMs": 10,
          "maxAllowedResponseTimeMs": 10
      }
    },
    {
      "name": "<entity-name>",
      "status": "Healthy",
      "tags": ["endpoint", "graphql"]
      "data": {
          "responseTimeMs": 20,
          "maxAllowedResponseTimeMs": 50
      }
    },
    {
      "name": "<entity-name>",
      "status": "Unhealthy",
      "tags": ["endpoint", "graphql"]
      "exception": "{exception-message-here}",
      "data": {
          "responseTimeMs": 20,
          "maxAllowedResponseTimeMs": 10
      }
    },
  }
}

> ```

Reviewer 4: Errors: 

1. "dab-configuration" is just "configuration"
2. checks is an array of objects, not an object of objects
3. "responseTimeMs" is "duraction-ms"
4. "maxallowedresponsetimems" is "threshold-ms"

Author: 1. Modified
2. Modified
3 and 4. Is this recently changed?
<img width="293" alt="image" src="https://github.com/user-attachments/assets/8e2d2bb8-8685-4756-8773-009ce40c9cc6" />

## Test Scenarios

Health Check result scenarios for different cases of Health check and GraphQL and Rest entities enabled or disabled.

> '_' means enabled/disabled (doesn't matter)

* Roles (Global Health and Entity Health are both enabled)
  * Runtime Health check parameter, `runtime.health.roles` contains "UserRole". However entity doesn't have read permissions for this role.

>     * Health for this entity is displayed with `status: Unhealthy` and `exception: Health could not be checked for this entity as it does not have permissions to perform read operation`. 

Reviewer 4: I believe it would better to allow for a 30x http status to be returned and react to that. This makes it so you do not have to look up entity permissions, you call every entity and simply handle them all the same. This enforces permissions without requiring the health check to look them up - which seems far more ideal.

    * Entity health DISABLED and Entity GraphQL ENABLED : GraphQL Health check `omitted` for this entity
    * Entity health _ and Entity GraphQL DISABLED : GraphQL Health check shows `status: Unhealthy` and `exception: Health could not be checked for this entity as it is disabled in config`. 
  
  * Global GraphQL Disabled
    * Entity health _ and Entity GraphQL _ : GraphQL Health checks are `omitted` from Health Report.
    
  * Global REST Enabled
    * Entity health ENABLED and Entity REST ENABLED : Health is shown for this particular entity with `status: Healthy`.
    * Entity health DISABLED and Entity REST ENABLED : REST Health check `omitted` for this entity
    * Entity health _ and Entity REST DISABLED : REST Health shows check `status: Unhealthy` and `exception: Health could not be check for this entity as it is disabled in config`. 
  
  * Global REST Disabled
    * Entity health _ and Entity REST _ : REST Health checks are `omitted` from Health Report.

* Global Health Disabled
  * Global GraphQL/REST _
    * Entity health _ and Entity GraphQL/REST _ : Original Health Format Report

## Limitations

+ We do not support health checks for stored procedures.

> + Hot-Reload is not supported in Comprehensive Health Endpoint.

Reviewer 4: Why?

Author: This needs to investigated further. Already tracking it

> + Multiple data-source configs are not supported in the HealthCheck Report

Reviewer 4: What!? Why? This may have been true when we were shooting for v1.4, but now that we are in 1.5, we have time to do this right.

Author: We have a separate issue to track this. If we get time, we can include this in 1.5. Have a separate PR to add the details about multiple data sources later in the design document

}
```

### Configuration Updates
The customer needs to update the `dab-config.json` file to include details of the health check for different configuration parameters like runtime, data source and entities. 

#### `runtime.health` Configuration
The runtime configuration would include details like `cache-ttl-seconds` in case we need to cache the response of health checks, the `max-dop` value which specifies the degree of parallelism i.e. how many queries that DAB should run in parallel and `roles` i.e. all roles which are allowed to view the comprehensive health information of DAB engine.

| **Property**   | **Data Type** | **Required** | **Default** | **Description**                                                                                      |
|----------------|---------------|--------------|-------------|------------------------------------------------------------------------------------------------------|
| `enabled`     | Boolean       | No           | `true`      | Enables or disables the comprehensive health checks for DAB Engine. **In case it is disabled, the '/health' updated health endpoint returns 404 (Not Found).**                                            |

> | `cache-ttl-seconds`   | Integer       | No           | `5`         | Time-to-live (in seconds) for caching health check results. **If this value is not specified or is 0, caching would not be enabled.** Currently, caching is not implemented hence, value is considered to be null.                                        |

Reviewer 4: > Currently, caching is not implemented hence, value is considered to be null.

This is unclear, it makes it sound like caching is not going to be part of the solution when it is. If you are not planning to implement caching, please let me know.

Author: Will be done subsequently. Not in current design scope.

| `max-dop`     | Integer       | No           | `8`         | Maximum Degree of Parallelism for running health checks.                                             |

> | `roles`       | Array of strings         | <table><tr><th>Environment</th><th>Required</th></tr><tr><td>Development</td><td>No</td></tr><tr><td>Production</td><td>Yes</td></tr></table>   | <table><tr><th>Environment</th><th>Default</th></tr><tr><td>Development</td><td>["anonymous"]</td></tr><tr><td>Production</td><td>NA</td></tr></table>        | List of Roles allowed to access the comprehensive health endpoint (e.g., `anonymous`, `authenticated` or `customer-role`).                   |

Reviewer 4: nit: customer-role > custom-role

```
{
  "runtime": {
    "health": {
      "enabled": true,
      "cache-ttl-seconds": 5,
      "max-dop": 5,
      "roles": ["anonymous", "authenticated"]
    }
  }
}
```

#### `data-source.health` Configuration

> The database type in the data source health config determines the query that we should run on the data source. The time to run this query should come under the user specified threshold_ms to qualify as a healthy data source for DAB. We get the database type from the runtime parameters to get the query to run on the specific DB Type.\

Reviewer 4: nit: threshold_ms > threshold-ms

> **This check opens a connection to the data source and validates the connection string and network access. Further it runs a simple query to validate user permissions and responsiveness or the DB.**

Reviewer 4: nit: responsiveness or the DB > responsiveness for the DB

| **Property**      | **Data Type** | **Required** | **Default** | **Description**                                                                                      |
|-------------------|---------------|--------------|-------------|------------------------------------------------------------------------------------------------------|
| `enabled`         | Boolean       | No           | `true`      | Enables or disables health checks for the data source.                                             |
| `name`         | String        | No           | Database Type `runtime.data-source.database-type`     | Identifier for the data source. Useful when multiple data sources exist, to identify each independently in the health check.  
| `threshold-ms`    | Integer       | No           | `10000`     | Threshold in milliseconds for the query response time before the check is considered unhealthy. Minimum : 1 Maximum: INT_MAX. Zero is not allowed.      |

```
{
  "data-source": {
    "health": {
      "name": "sqlserver",
      "enabled": true,
      "threshold-ms": 100
    }
  }
}
```

#### `<entity-name>.health` Configuration

> The Entity config parameters contain information about the `first` which defines the number of first rows to be returned when executing a READ operation on the data source entity and under what threshold ms should the response be received for it to qualify as a healthy entity.

Reviewer 4: nit: parameters > settings
nit: defines the number of first rows > defines the number of rows
nit: under what threshold ms should the response be received > under what threshold the response should be received

| **Property**     | **Data Type** | **Required** | **Default** | **Description**                                                                                      |
|------------------|---------------|--------------|-------------|------------------------------------------------------------------------------------------------------|
| `enabled`        | Boolean       | No           | `false`      | Enables or disables health checks for the specific entity.                                          |
| `first`          | Integer       | No           | `100`         | Number of records to query during the health check. Value is used to create the SELECT query to fetch records for REST and GraphQL endpoints. Minimum : 1 Maximum: INT_MAX. Zero is not allowed.                                                |

> | `threshold-ms`   | Integer       | No           | `10000`     | Threshold in milliseconds for the query response time before the check is considered degraded. Minimum : 1 Maximum: INT_MAX. Zero is not allowed.     |

Reviewer 4: > considered degraded

We are not using the health status `degraded` just `healthy` and `unhealthy`.

        "first": 1 (optional; default: 100),
        "threshold-ms": 100 (optional; default: 10000)
      },
      ...
    },
  }
}
```
The idea of using this updated configuration is to allow the developer to influence how the health checks work against the datasource/entity. This would provide them with a more detailed process for checking if DAB engine is healthy and would give them an enhanced user experience. 

### Permissions for Health Check Report (Roles)
We focus on two aspects in terms of roles for health report. 

+ Health Report Access\
We validate if the incoming user has access to the comprehensive health check report i.e. check the allowed roles in the `runtime.health.roles` array in the config. Only if the user has access, we display the health report, else we show a 403 Forbidden error message.
+ Read Permissions Access\
We check if this incoming role is allowed to perform the `read` query on the DB. For this we focus on the entity section in the config, where the incoming role should be added with `read` permissions. Else, the http client trying to run the `read` command will fail with 403. We return the error message from http client with Unhealthy status. 

> > For data source health check we only need to check health report access but for entities (rest and graphql) we run both health report access and read permissions access. 

Reviewer 4: Unclear. Will this be true for Health Checks or only Static Web Apps?

There are three kinds of roles in DAB.\
System roles are built-in roles recognized by Data API builder. A system role is auto assigned to a requestor regardless of the requestor's role membership denoted in their access tokens. There are two system roles: anonymous and authenticated.
+ Anonymous\
This is the superset of all roles. The anonymous system role is assigned to requests executed by unauthenticated users. Runtime configuration defined entities must include permissions for the anonymous role if unauthenticated access is desired.
+ Authenticated\
The authenticated system role is assigned to requests executed by authenticated users. All custom roles are part of Authenticated role. 
+ Custom Roles\
Custom roles are non-system roles that are assigned to users within the identity provider you set in the runtime config. 

> In Static Web Apps, a user is a member of the anonymous role by default. If the user is authenticated, the user is a member of both the anonymous and authenticated roles. If the client application's request includes the HTTP header `X-MS-API-ROLE` with a custom value, the request is evaluated in the context of the 'custom role'. 

> **Access to comprehensive health report is measured via the `runtime.health.roles` array where `anonymous` is added to allow unauthenticated access to an entity, `authenticated` is added to allow all custom roles, and `custome-role` is added to allow that specific role to view the report.**

Reviewer 4: nit: custome-role > custom-role

### Base URL for DAB
While running the health checks for entities, we need the BASE URL on which DAB is running. This BASE URL is used to run all datasource, rest and graphql health check queries for our engine. This base URL is identified during runtime by using `httpcontext`. The following command fetches the base URL upon which all health checks would be run subsequently. 
```
string path = UriHelper.GetEncodedUrl(httpContext!.Request).Split('?')[0];
```

## Health Endpoint Report Calculation 
This section contains details on how the health check details are calculated and Comprehensive Health Check report is formulated at the newly updated endpoint: `/health`.

> Case 1: `runtime.health.enabled` is false/null\
The endpoint returns a 404 (Not Found) error response when hitting the page as comprehensive health check report is disabled.

> Case2:  `runtime.health.enabled` is true\
The endpoint returns the comprehensive health check report after executing the checks for data-source and entities.

### Configuration Parameters

> The Health Report also contains a section for the configuration summary for the health checks. It provides information about various parameters being enabled or disabled on the global runtime scale. 

Reviewer 4: nit: parameters > settings

> The Health Report also contains a section for the configuration summary for the health checks. It provides information about various parameters being enabled or disabled on the global runtime scale. 

Reviewer 4: > global runtime scale

Scale? Not sure what that is supposed to mean.

It includes the following
+ Http
+ Https
+ Rest
+ GraphQL
+ Telemetry
+ Caching
+ Mode of Deployment

> NOTE that BASIC health does NOT include configuration.

### Application Name
The application name created by DAB that is appended to data source connection string; is a rich string that exposes a lot of internal stuff. This would be enhanced in the future, hence use the `GetDataApiBuilderUserAgent()` function from the `ProductInfo` class to fetch this for the health report. 

### DataSource Health Check

> DAB Config has two parameters in the `data-source.health` config section. (Enabled and threshold_ms)\

Reviewer 4: nit: two parameters > two settings

Just a reminder, a `parameter` is a variable listed in a method’s signature, used to pass data into the method.

> DAB Config has two parameters in the `data-source.health` config section. (Enabled and threshold_ms)\

Reviewer 4: nit: threshold_ms > threshold-ms

For each database, we execute a standard query (given below) to get the time elapsed result (ms) for that DB and use this value to compare with threshold according to the [Overall Health Calculation](#overall-health-calculation) and get the status as `Healthy/Unhealthy`.

The standard queries that would be run on each Database Type are the following
+ Postgres: SELECT 1;
+ MySQL: SELECT 1;
+ Cosmos DB: SELECT VALUE 1; (for SQL API)
+ MS SQL: SELECT 1;

**Here we add the tag `database` in the tags array to differentiate the datasource health check result from the entity health check result where `endpoint` is being added to Tags array.**

```
{
  "checks": [
    {
      "name": "<database-type>",
      "status": "Healthy/Unhealthy",

>       "tags": ["database"],

Reviewer 4: We will also add `database-type` as a tag.

To form the query we run a for loop against the entities in the config file. For each entity, we add two objects in the output checks array; first for REST and second for GraphQL. These objects are differentiated using the Tags array in the output report. For each rest or graphql we add the appropriate tag in the array for filtering. 
```
{
  "checks": [
    {
      "name": "<entity-name>",
      "status": "Healthy/Unhealthy",
      "tags": ["rest/graphql", "endpoint"],
      "data": {
        "duration-ms": , // Response time of query in ms
        "threshold-ms": , // Max Allowed response time for query in ms
      }
    }
  ]
}
```

#### Rest Query
We take the base URL and append the REST path in the suffix to get `<baseUrl>/<runtime.rest.path>`. (Default Rest Path: /api)\
For each entity, we need to build the rest query url using the above path as prefix. We build the entity REST path using `<entity-name>.rest.path ?? entityName`. This means if the path is given use that, else the entity name is the path to append to suffix.
Further our call becomes `<baseUrl>/<runtime.rest.path>/<entity-name>.rest.path`.

> Finally we add the query parameter which is the first 'x' values to get the final query as **`<baseUrl>/<runtime.rest.path>/<entity-name>.rest.path?$first=x`**.

Author: Here it would be a parameter itself as I am refering to query parameters not config

For this we run a **POST** query against the base URL `<baseUrl>/<runtime.graphql.path>` with a Request BODY.\
To create the request body we need to fetch the primitive column names for this entity. 

##### **Fetch Column Names**
To get column names for an entity, we use the 'MetadataProvider' for each DB Type. Each Metadata Provider has `EntityToDatabaseObject` which can be used to write the function `GetSchemaGraphQLFieldNamesForEntityName()` which provides the primitive column names for each entity.\
The `EntityToDatabaseObject` is a Dictionary with key as string containing the name of entity and value as DatabaseObject containing the metadata of the entity. This DatabaseObject contains a parameter `SourceDefinition` which holds information about the database table definition. It contains properties that are common between a database table and a view. This SourceDefinition contains a parameter `Dictionary<string, ColumnDefinition> Columns` which has key as column name and value as basic details of the column. 

**GraphQL Query**\
After getting column names array (`columnNames`) for the entity we create the graphQL query payload 
```

> query = $"{{{entityName.ToLowerInvariant()}(first: {First}) {{items {{ {string.Join(" ", columnNames)} }}}}}}"

Author: yes, this works

      "status": "Healthy",
      "tags": ["data-source", "mssql"],
      "data": {
          "duration-ms": 10,
          "threshold-ms": 100
      }
    },
    {
      "name": "<entity-name>",
      "status": "Healthy",
      "tags": ["endpoint", "rest"],
      "data": {
          "duration-ms": 2,
          "threshold-ms": 10
      }
    },
    {
      "name": "<entity-name>",
      "status": "Healthy",
      "tags": ["endpoint", "graphql"]
      "data": {
          "duration-ms": 20,
          "threshold-ms": 50
      }
    },
    {
      "name": "<entity-name>",
      "status": "Unhealthy",
      "tags": ["endpoint", "rest"]
      "exception": "{exception-message-here}",
      "data": {
          "duration-ms": 20,
          "threshold-ms": 20
      }
    },
    {
      "name": "<entity-name>",
      "status": "Unhealthy",
      "tags": ["endpoint", "graphql"]
      "exception": "{exception-message-here}",
      "data": {
          "duration-ms": 20,
          "threshold-ms": 10
      }
    },
  ]
}
```

## Test Scenarios

Health Check result scenarios for different cases of Health checks for DataSource, Rest and GraphQL entities to be enabled or disabled.

> ![Class Diagram](dataSourceHealthCheckTestScenarioFlowchart.png)

Reviewer 3: Thank you for adding these diagrams! Helps get an overall idea of the different conditions! 💯

| **Status.Unhealthy**| The system has a critical failure or issue.            |

## Proposed Change
We want to create a detailed version of the `Health endpoint for DAB Engine` with information regarding all REST and GraphQL endpoints and their behavior. Here we would update the configuration schema to include some properties which would be required from the user to perform all checks on that data-source/entity (described below) to validate if they are healthy.\
`We would have the basic health results displayed at the '/' (root) and create a new endpoint '/health' for the comprehensive health report depending on a mixture of if health is enabled and the user meets the roles requirement.`\
We want the user to provide a check and threshold value for each entity for which health is enabled and DAB engine would carry out this check. If the result is under the threshold, the DAB Engine would be considered `healthy` for that endpoint, else `unhealthy`.

### **Overall Health Calculation**
| **Healthy** | **Unhealthy** | **Global Status** |
|------------|--------------|-------------------|
| -          | 0            | `Healthy`        |

> | -          | ≥ 1          | `Unhealthy`      |

Reviewer 3: For any reason if an entity check is **skipped**, because rest is disable or graphql is disabled or role doesn't have permission, what do we consider it as being Healthy or Unhealthy for the overall health calculation? i.e. would something being skipped make the overall status - unhealthy?

Author: if the entity check is skipped it is not counted towards the overall health status calculation. Its ignored

| **Status.Unhealthy**| The system has a critical failure or issue.            |

## Proposed Change
We want to create a detailed version of the DAB Engine Health endpoint with information regarding all REST and GraphQL endpoints and their behavior. Here we would update the configuration schema to include what all results or checks do we want to perform on that data-source/entity (described below) to validate if they are healthy.
We want the user to provide a check and threshold value what DAB would carry out and then check if the result is under the threshold, if so, the DAB Engine would be considered healthy for that endpoint, else unhealthy. 

**Case 1: Datasource**\

> Here we would carry out standard queries based on the datasource type and execute them under the given threshold. If the engine gets the result below the specified threshold, the DB is considered healthy else unhealthy.

Reviewer 3: Are we no longer letting the health check query to be configured?

|----------------|---------------|--------------|-------------|------------------------------------------------------------------------------------------------------|
| `enabled`     | Boolean       | No           | `false`      | Enables or disables the comprehensive health checks for DAB Engine. In case of disabled; we show the previous format of health check.                                             |
| `cache-ttl-sec`   | Integer       | No           | `5`         | Time-to-live (in seconds) for caching health check results. If this value is not specified, caching would not be enabled. Currently, caching is not implemented hence, value is considered to be null.                                        |
| `max-dop`     | Integer       | No           | `1`         | Maximum Degree of Parallelism for running health checks.                                             |
| `roles`       | Array of strings         | Yes           | NA         | Roles allowed to access the health endpoint (e.g., `anonymous`, `authenticated`).                   |

#### `data-source.health` Configuration

> The database type in the data source health config determine the threshold of ms it should come under to qualify as a healthy data source for DAB. We get the database type from the runtime parameters to get the query to run on the specific DB Type.

Reviewer 3: ```suggestion
The database type in the data source health config determines the query that we should run on the data source. The time to run this query should come under the user specified threshold of ms to qualify as a healthy data source for DAB. We get the database type from the runtime parameters to get the query to run on the specific DB Type.
```

 > TODO: Handle Health Endpoint for multiple data-source configs in the upcoming enhancements

| **Property**      | **Data Type** | **Required** | **Default** | **Description**                                                                                      |
|-------------------|---------------|--------------|-------------|------------------------------------------------------------------------------------------------------|
| `enabled`         | Boolean       | No           | `true`      | Enables or disables health checks for the data source.                                              |

> | `name`         | Array of String        | No (Yes in case of multiple data source)           | Database Type      | Identifier for the data source; useful when multiple data sources exist.  

Reviewer 3: is the name really an Array of string?

Author: yes this field would be elaborated once we consider the case of multiple data sources. Hence array. Else it would not be needed as db type is taken

#### `runtime.health` Configuration

The runtime configuration would include details like cache-ttl in case we need to cache the response of health checks, the max-dop value which specifies the degree of parallelism i.e. how many queries that DAB should run at once to get health results and roles i.e. which role is allowed to view the health information of DAB.

| **Property**   | **Data Type** | **Required** | **Default** | **Description**                                                                                      |
|----------------|---------------|--------------|-------------|------------------------------------------------------------------------------------------------------|

> | `enabled`     | Boolean       | No           | `false`      | Enables or disables the comprehensive health checks for DAB Engine.                                             |

Reviewer 3: When this is enabled = false, what do we show on the /health page?

Reviewer 3: So, what is shown on this `/health` page in the current way? 

Also, what if someone has an entity named `/health` - wont enabling the new endpoint conflict to that entity's page? What do we do if there is such a conflict?

Author: We didnt discuss this in the design discussion. Let me put this question to Jerry.
@JerryNixon Please share your thoughts.

> | `cache-ttl`   | Integer       | No           | `5`         | Time-to-live (in seconds) for caching health check results. If this value is not specified, caching would not be enabled.                                          |

Reviewer 3: can we add the unit of time in the property name? like `cache-ttl-sec`

| `max-dop`     | Integer       | No           | `1`         | Maximum Degree of Parallelism for running health checks.                                             |
| `roles`       | Array of strings         | Yes           | NA         | Roles allowed to access the health endpoint (e.g., `anonymous`, `authenticated`).                   |

#### `data-source.health` Configuration

> The data source config parameters specify the query that we should run on the data source and the threshold of ms it should come under to qualify as a healthy data source for DAB.

Reviewer 3: If we removed the query, why does this section still say that the `query` will be specified?

| **Property**      | **Data Type** | **Required** | **Default** | **Description**                                                                                      |
|-------------------|---------------|--------------|-------------|------------------------------------------------------------------------------------------------------|
| `enabled`         | Boolean       | No           | `true`      | Enables or disables health checks for the data source.                                              |
| `name`         | String        | No (Yes in case of multiple data source)           | Database Type      | Identifier for the data source; useful when multiple data sources exist.  
| `threshold-ms`    | Integer       | No           | `10000`     | Threshold in milliseconds for the query response time before the check is considered degraded.      |

#### `<entity-name>.health` Configuration

> The Entity config parameters contain information about the GET filter which needs to be carried out on the data source entity and under what threshold ms should the response be received for it to qualify as a healthy entity.

Reviewer 3: filter is removed. So, the description should be updated accordingly.

Reviewer 3: Again, I don't see the update in the latest iteration. I cannot sign off unless I see it here.

Author: <img width="591" alt="image" src="https://github.com/user-attachments/assets/e362dacc-245f-4fc1-9671-1eddd2b8611f" />
The changes are already updated. 
<img width="689" alt="image" src="https://github.com/user-attachments/assets/b9873f1b-404d-4e57-922f-997994d9841e" />

| `filter`         | String        | No           | `null`      | Filter condition applied to the health check query (e.g., `"Id eq 1"`).                             |
| `first`          | Integer       | No           | `1`         | Number of records to query during the health check.                                                 |
| `enabled`        | Boolean       | No           | `false`      | Enables or disables health checks for the specific entity.                                          |
| `first`          | Integer       | No           | `1`         | Number of records to query during the health check. Using this value the GET query would be created to fetch records for REST and GraphQL endpoints.                                                 |
| `threshold-ms`   | Integer       | No           | `10000`     | Threshold in milliseconds for the query response time before the check is considered degraded.      |

#### Example
```
{
  "runtime" : {
    ...
    "health" : {
      "enabled": true, (default: true)

>       "cache-ttl": 5, (optional default: 5)

Reviewer 3: What is the `default` if cache-ttl is optional?

Author: Default would is null currently. But after caching is implemented it would be 5 sec

+ NOTE that BASIC health does NOT include configuration.

### DataSource Health Check
For each database we have two parameters in the health check config parameters. (Enabled and threshold)\
For each database type we execute a standard query to get the results for that DB and check if the time elapsed is under the threshold.
The standard queries that would be run on each Database Type are the following
+ Postgres: SELECT 1 LIMIT 1;
+ MySQL: SELECT 1 LIMIT 1;
+ Cosmos DB: SELECT TOP 1 1 FROM c
+ MS SQL: SELECT TOP 1 1

### Entity Health Check

In the config, for each entity we have three health parameters (Enabled, First and Threshold) 

+ Enabled: This would define whether we need to run the comprehensive health check for this role
+ First: This would be used to create the query for endpoints to fetch first 'x' records for the DB.
+ Threshold: This is used to check the time elapsed to run the above query.

> To form the query we run a for loop against the entities in the config file. For each entity, we add two parameters in the checks array of the output. One for REST and another one for GraphQL, these calls are differentiated using the Tags array in the output report.

Reviewer 3: if an entity is disabled for `rest`, but its health check is `enabled`  - do we still show the health check results? Same for graphql

If we a role doesn't have `read` permissions on the entity, but its health check is enabled for that role, do we display the health check results for that entity?

Author: entity is disabled for rest, but its health check is enabled 
We would show an error in the comprehensive health check saying "Unhealthy" with the exception message saying "Entity is disabled in Config"

 role doesn't have read permissions on the entity, but its health check is enabled for that role
We would show an error in the comprehensive health check saying "Unhealthy" with the exception message saying "Role is not permitted to perform the SELECT query".

Reviewer 3: `Role is not permitted to perform the SELECT query` --> `Role is not permitted to perform the read operation.`

Please confirm this behavior and the respective error message with @JerryNixon

Author: If you see the test scenarios section. I have written the second part only with read operation not SELECT query

Author: I discussed it with the team and Jerry on call.  
@JerryNixon , Please confirm if this is the expected behavior

Author: updated as per new

#### Rest Query
We take the base URL which is http://localhost:5000 and append the REST path in the suffix. After this we get http://localhost:5000/api.\
For each entity we have rest as a parameter in the config block as well in case the path of entity is different from its name. We build the entity REST path using `<entity-name>.rest.path ?? entityName`. This means if the path is given use that, else the entity name is the path to append to suffix.
Further our call becomes ` http://localhost:5000/api/usertable`.

Finally we add the query parameters which is first to get the query to execute as `http://localhost:5000/api/usertable?$first=5`. This query is then executed to get the elapsed time and we get whether the REST endpoint of UserTable healthy or not.

#### GraphQL Query

> To execute this query we first need to get the schema of that particular table. Hence for each table we run the `introspection query` to get the schema which gives the column names which are then used to query the GraphQL endpoint to check the health.

Reviewer 3: ```suggestion
To execute this query we first need to get the schema of that particular entity. Hence for each entity we run the `introspection query` to get the schema which gives the field names which are then used to query the GraphQL endpoint to check the health.
```

> For this we run a **POST** query against the base URL **http://localhost:5000/graphql/** with a Request BODY.\

Reviewer 3: what if the base url is different than http://localhost:5000? will this work when dab is running in a docker container?

Author: For the current case, I have focused on keeping URL as this. Will open an issue to track in case URL is different

Reviewer 3: Can't you determine the base url in code without hardcoding localhost?

Author: Can you suggest some way

Reviewer 3: The httpContext request to the health endpoint should have the compelete url from which you can derive the base url..

See this code in the repo:
   `string path = UriHelper.GetEncodedUrl(httpContext!.Request).Split('?')[0];`

Author: done

The CURL Command of that query is 
```
curl --request POST \
  --url http://localhost:5000/graphql \
  --header 'Content-Type: application/json' \
  --header 'User-Agent: insomnia/10.3.0' \
  --data '{
  "operationName": "IntrospectionQuery",

>   "query": "query IntrospectionQuery {\n  __schema {\n    queryType {\n      name\n    }\n    mutationType {\n      name\n    }\n    subscriptionType {\n      name\n    }\n    types {\n      ...FullType\n    }\n    directives {\n      name\n      description\n      isRepeatable\n      args {\n        ...InputValue\n      }\n      locations\n    }\n  }\n}\n\nfragment FullType on __Type {\n  kind\n  name\n  description\n  specifiedByURL\n  oneOf\n  fields(includeDeprecated: true) {\n    name\n    description\n    args {\n      ...InputValue\n    }\n    type {\n      ...TypeRef\n    }\n    isDeprecated\n    deprecationReason\n  }\n  inputFields {\n    ...InputValue\n  }\n  interfaces {\n    ...TypeRef\n  }\n  enumValues(includeDeprecated: true) {\n    name\n    description\n    isDeprecated\n    deprecationReason\n  }\n  possibleTypes {\n    ...TypeRef\n  }\n}\n\nfragment InputValue on __InputValue {\n  name\n  description\n  type {\n    ...TypeRef\n  }\n  defaultValue\n}\n\nfragment TypeRef on __Type {\n  kind\n  name\n  ofType {\n    kind\n    name\n    ofType {\n      kind\n      name\n      ofType {\n        kind\n        name\n        ofType {\n          kind\n          name\n          ofType {\n            kind\n            name\n          }\n        }\n      }\n    }\n  }\n}"

Reviewer 3: would be helpful to see a more readable version of the Introspection query

}'
```
**Introspection Query Filter**

This gives us the GraphQL Schema which is deserialized to get the column names.\
Now we need to create a `columnNames` array which contains the list of column names which can be used to create the graphQL query.\
In the output of the above CURL command we get `Data.Schema.Types` array which contains these column names. We loop through this array and Match the `type.kind` value with `OBJECT` and `type.name` value with the name of the entity `UserTable`. (Only one entry satisfies this check)

> After identifying that specific entity `type` that satisfies the above conditions, we get the object that contains the entity column names. However an entity can have primitive column types or Object column type which are nested further. To resolve this we loop on `type.fields` array. For each field we get the `field.type` object and check if the `field.type.kind` is `SCALAR` then this is a primitive column name and I add it to the `columnName` array above or if the `field.type.kind` is `NON_NULL` then `field.type.ofType.kind` should be `SCALAR`. We add all those column names which satisfies either of these two conditions to the top `columnNames` array and finally use this to create the graphQL query.

Reviewer 3: when `field.type.kind` is SCALAR, isnt it NON_NULL as well? If yes. I dont see how these 2 conditions are different. What is the oher value of `field.type.kind` that we will then look for `field.type.ofType.kind`

Reviewer 3: Maybe my question isn't clear.. 

Wouldn't the condition field.type.kind is NON_NULL be true when field.type.kind is SCALAR?

Is your logic to check for Case 2 only when Case 1 is FALSE?

Author: yes in the if condition, the case 1 is checked first. Only when case 1 is false, the other case 2 is checked. 
In case case 1 is positive, Having and OR condition would make the if condition true without evaluation case 2.

Also let me elaborate, Having type.kind as NON_NULL is not alone a valid scenario. Along with this, it needs to have type.ofType.kind as SCALAR as well. Only then case 2 is considered positive.

+ First: This would be used to create the query for endpoints to fetch first 'x' records for the DB.
+ Threshold: This is used to check the time elapsed to run the above query.

To form the query we run a for loop against the entities in the config file. For each entity, we add two parameters in the checks array of the output. One for REST and second for GraphQL, these calls are differentiated using the Tags array in the output report.\

While running the health checks for entities, we need the BASE URL on which DAB is running which we use to run all rest and graphql queries on our engine. 
> Note: Current implementation we consider `http://localhost:5000` as the base URL. Need to track this, in case different. Everywhere in the document, using `baseUrl` in place of `http://localhost:5000`.

#### Rest Query

> We take the base URL which is and append the REST path in the suffix. After this we get baseUrl/api. (Default Rest Path: /api)\

Reviewer 3: ```suggestion
We take the base URL and append the REST path in the suffix. After this we get baseUrl/api. (Default Rest Path: /api)\
```

Now we need to create a `columnNames` array which contains the list of column names which can be used to create the graphQL query.\
In the output of the above CURL command we get `Data.Schema.Types` array which contains these column names. We loop through this array and match the `type.kind` value with `OBJECT` and `type.name` value with the name of the entity i.e. `UserTable`. (Note: Only one entry satisfies this check)
After identifying that specific entity `type` that satisfies the above conditions, we get the object that contains the entity column names. However an entity can have primitive column types or Object column type which are nested further. These object nested types cannot be used to create the graphql query and hence we need to filter out the primitive data types. To resolve this we loop on `type.fields` array and match it with two conditions. If either of the conditions match, we add it to the column names array.\
For each field we get the `field.type` object 
+ Condition 1: If the `field.type.kind` is `SCALAR` then this is a primitive column name.
+ Condition 2: If the `field.type.kind` is `NON_NULL` then check if `field.type.ofType.kind` is `SCALAR`. 

> We add all those column names which satisfies either of these two conditions to the top `columnNames` array and finally use this to create the graphQL query.

Reviewer 3: ```suggestion
We add all those column names which satisfy either of these two conditions to the top `columnNames` array and finally use this to create the graphQL query.
```

We focus on two aspects in terms of roles for health report. 

+ Health Report Access\
We validate if the incoming user has access to the comprehensive health check report i.e. check the allowed roles in the `runtime.health.roles` array in the config. Only if the user has access, we display the health report, else we show a 403 Forbidden error message.
+ Read Permissions Access\
We check if this incoming role is allowed to perform the `read` query on the DB. For this we focus on the entity section in the config, where the incoming role should be added with `read` permissions. Else, the http client trying to run the `read` command will fail with 403. We return the error message from http client with Unhealthy status. 

> For data source health check we only need to check health report access but for entities (rest and graphql) we run both health report access and read permissions access. 

There are three kinds of roles in DAB.\
System roles are built-in roles recognized by Data API builder. A system role is auto assigned to a requestor regardless of the requestor's role membership denoted in their access tokens. There are two system roles: anonymous and authenticated.
+ Anonymous\

> This is the superset of all roles. The anonymous system role is assigned to requests executed by unauthenticated users. Runtime configuration defined entities must include permissions for the anonymous role if unauthenticated access is desired.

Reviewer 3: technically anonymous is not "superset" since an authenticated request is not anonymous. If an entity has read permission granted to authenticated only, then anonymous cant perform read. In that sense, anonymous is not superset.

Author: Superset would mean that if permission is granted to the parent its accessible to all children. But it doesnt mean that if permission is granted to child, then it flows to parent. 
Superset here is reffered as if health has roles array. And if this roles array has anonymous role, then everyone can access it. We need not check again more. 
Updated to more clear

+ Authenticated\
The authenticated system role is assigned to requests executed by authenticated users. All custom roles are part of Authenticated role. 
+ Custom Roles\
Custom roles are non-system roles that are assigned to users within the identity provider you set in the runtime config. 

> > Example: In Static Web Apps, a user is a member of the anonymous role by default. If the user is authenticated, the user is a member of both the anonymous and authenticated roles. If the client application's request includes the HTTP header `X-MS-API-ROLE` with a custom value, the request is evaluated in the context of the 'custom role'. 

Reviewer 3: the user may be a member of both the roles - anonymous and authenticated. But, I think we evaluate the request in the context of `authenticated` role only.

Author: yes that is correct. Added specific text for this

    "graphql": true,
    "telemetry": true,
    "caching": true,
    "mode": "development"
  },
  "checks": {
    {
      "name": "database-name",
      "status": "Healthy",
      "tags": ["data-source", "mssql"],
      "data": {
          "responseTimeMs": 10,
          "maxAllowedResponseTimeMs": 10
      }
    },
    {
      "name": "<entity-name>",
      "status": "Healthy",
      "tags": ["endpoint", "rest"],
      "data": {
          "responseTimeMs": 10,
          "maxAllowedResponseTimeMs": 10
      }
    },
    {
      "name": "<entity-name>",
      "status": "Healthy",
      "tags": ["endpoint", "graphql"]
      "data": {
          "responseTimeMs": 20,
          "maxAllowedResponseTimeMs": 50
      }
    },
    {
      "name": "<entity-name>",
      "status": "Unhealthy",
      "tags": ["endpoint", "graphql"]
      "exception": "{exception-message-here}",
      "data": {
          "responseTimeMs": 20,
          "maxAllowedResponseTimeMs": 10
      }
    },
  }
}
```

## Test Scenarios

Health Check result scenarios for different cases of Health check and GraphQL and Rest entities enabled or disabled.\
'_' means enabled/disabled (doesn't matter)

> * Global Health Enabled

Reviewer 3: Good to enlist test scenarios and expectations of when role is specified and entity has read permissions for that role or not.

Author: added another test scenario on roles

## Test Scenarios

Health Check result scenarios for different cases of Health check and GraphQL and Rest entities enabled or disabled.

> '_' means enabled/disabled (doesn't matter)

* Roles (Global Health and Entity Health are both enabled)

>   * Runtime Health check parameter, `runtime.health.roles` contains "UserRole". However entity permissions doesn't have read permissions on this entity

Reviewer 3: ```suggestion
  * Runtime Health check parameter, `runtime.health.roles` contains "UserRole". However, the `UserRole` doesn't have read permissions on this entity.
```

    * Health for this entity is displayed with `status: Unhealthy` and `exception: Health could not be checked for this entity as it does not have permissions to perform read operation`. 
  
**Cases where Global or Entity health and REST and GraphQL for that Entity is enabled or disabled**
* Global Health Enabled
  * Global GraphQL Enabled
    * Entity health ENABLED and Entity GraphQL ENABLED : Health is shown for this particular entity with `status: Healthy`.
    * Entity health DISABLED and Entity GraphQL ENABLED : GraphQL Health check `omitted` for this entity

>     * Entity health _ and Entity GraphQL DISABLED : GraphQL Health check shows `status: Unhealthy` and `exception: Health could not be check for this entity as it is disabled in config`. 

Reviewer 3: ```suggestion
    * Entity health _ and Entity GraphQL DISABLED : GraphQL Health check shows `status: Unhealthy` and `exception: Health could not be checked for this entity as it is disabled in config`. 
```
